### PR TITLE
Search backend: use efficient implementation for repo:contains.file()

### PR DESCRIPTION
**Pull request by camdencheek**
This just updates our `repo:contains.file()` predicate to hook into the
more efficient machinery we already have for `repohasfile`.

Note that this does not cover the `repo:contains(file:xyz)` case since
we do not currently have an efficient implementation of a "repo contains content" 
filter.

## Test plan

Added plan tests to show that we generate the same `TextPatternInfo` with `repohasfile`
and `repo:contains.file()`

### CODE REVIEW

File: internal/search/query/predicate.go [hunk 1]
```diff
@@ -237,8 +237,8 @@ func (f *RepoContainsFilePredicate) ParseParams(params string) error {
 func (f *RepoContainsFilePredicate) Field() string { return FieldRepo }
 func (f *RepoContainsFilePredicate) Name() string  { return "contains.file" }
 func (f *RepoContainsFilePredicate) Plan(parent Basic) (Plan, error) {
-	contains := RepoContainsPredicate{File: f.Pattern, Content: ""}
-	return contains.Plan(parent)
+	// Handled by repo search
+	return nil, nil
```
Comment (camdencheek) [inline internal/search/query/predicate.go:241]: We have this slightly awkward assumption in the design of predicates that all predicates will be expanded in advance of executing the query (thanks me). 

To get around that, we just return a nil plan, which our machinery will interpret as "don't expand this" then the predicate is handled by repo search.

File: internal/search/query/types.go [hunk 2]
```diff
@@ -330,6 +330,41 @@ func (p Parameters) IncludeExcludeValues(field string) (include, exclude []strin
...
+		if name != "contains.file" {
+			return
+		}
+
+		var p RepoContainsFilePredicate
+		if err := p.ParseParams(params); err != nil {
+			return
+		}
+
+		if negated {
```
Comment (camdencheek) [inline internal/search/query/types.go:358]: Note that we don't currently allow predicates to be negated, but I think we can remove that blanket requirement because this can, in fact, be executed efficiently.

File: internal/search/query/types.go [hunk 3]
```diff
@@ -330,6 +330,41 @@ func (p Parameters) IncludeExcludeValues(field string) (include, exclude []strin
...
+
+		name, params := ParseAsPredicate(v)
+		if name != "contains.file" {
+			return
+		}
+
+		var p RepoContainsFilePredicate
+		if err := p.ParseParams(params); err != nil {
+			return
+		}
```
Comment (camdencheek) [inline internal/search/query/types.go:356]: This could probably be abstracted better, but I don't think it's too bad until we figure out some better patterns here.

**Review by tbliu98 — APPROVED** [review]

File: internal/search/query/predicate.go (hunk 1 above)
Comment (camdencheek) [reply internal/search/query/predicate.go:241]: The issue we face in predicates is actually a pretty common issue in query engines of all types. Basically, if you have some set of filters, there are a few types of strategies you can use to calculate the intersection. I'm going to overexplain a bit because I'm finding it useful to organize my thoughts here as well.

(1) The naive way is to collect the full list of things that match one filter, then collect the full list of things that match the second filter, then run the intersection. This is usually quite expensive because the intersection might be quite small compared to the size of each of the intermediate sets. This is the behavior of `AndJob`, which acts as a catch-all "handle an intersection when we have no better way to do it."

(2) A different and sometimes better method is to generate the list of things that match the first filter, then "push down" that list to the backend that handles the second filter. This is useful when you have a backend that knows how to handle the second filter, but not the first filter. This the case for most predicates: our backends all know how to efficiently filter by repo name and file content, but they don't all know how to natively handle `repo:contains.file()`. So, we expand `repo:contains.file()` into a set of repo names, which our backends know how to handle. This method works well when the expanded set is small, which is often the case for repo predicates.

(3) Another similar approach is to query the backends with only the filters that it can handle, then post-filter the results with the filters it can't handle. For example, if instead of expanding `repo:contains.file(Dockerfile)` to a set of repos, then sending it to the backend, we'd just query the backend for results that match the other filters, then for each result that comes through, check to see if its repo contains `Dockerfile`. This approach works well when we expect the expanded set to be large, but the set returned by the backend to be small. Think of a query like `repo:contains.file(\.go$) super_rare_string_that_will_only_have_one_match`. It's way more efficient to just search for the string, then post-filter than it would be to expand `repo:contains.file(\.go$)` and send that massive list to Zoekt. We use this approach with the `select:` filter because we run the search as normal without `select:`, then post-process the results and deduplicate.

(4) The holy grail is when all filters can be "pushed down" to the backends and the backends themselves can efficiently handle the filter. (I put "pushed down" in quotes because this is a term you'll see in DB engine literature). We do this for and/or/not queries with Zoekt (instead of joining them with an inefficient `AndJob`), but Zoekt doesn't support `repo:contains.file()`, so we have to handle that in the query engine. 

The implementation of predicates by me really only had (2) in mind. And actually, this PR doesn't change that we're still doing (2). However, the _big_ problem with predicates is their interface forces you to expand them _into another query_. This means that `repo:contains.file(Dockerfile) test`  will first be expanded to the actual query `(repo:^github\.com/sourcegraph/sourcegraph$|...|^github.com/sourcegraph/zoekt$) test`.

This is really problematic because (a) our query processing wasn't built for the kind of massive queries this generates, and it's really inefficient, and (b) **we can only create predicates that are representable as a search query**. With this constraint, it would make it impossible to implement something like `repo:description(test)` because you can't write a search query that searches by description. This bites us right now because `repo:contains.file(xyz)` [expands to](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@c11b39309c57e9ef746b8cea6e15b214eb46affb/-/blob/internal/search/query/predicate.go?L170-190) `file:xyz select:repo`, which searches for every file that matches `xyz`, then post-filters with the select (approach (3) above). 

TL;DR: we want flexibility to take alternative approaches to pushing down filters, and the constraint that a predicate has to be expandable to a search query is bad.

Comment (tbliu98) [reply internal/search/query/predicate.go:241]: I genuinely appreciate over-explanations, so tyvm


> This method works well when the expanded set is small, which is often the case for repo predicates.

Do we have data on this?

File: internal/search/query/types.go (hunk 3 above)
Comment (rvantonder) [reply internal/search/query/types.go:356]: yeah we can figure this out lator

File: internal/search/query/types.go [hunk 4]
```diff
@@ -330,6 +330,41 @@ func (p Parameters) IncludeExcludeValues(field string) (include, exclude []strin
...
+
+	VisitField(nodes, FieldRepo, func(v string, negated bool, ann Annotation) {
+		if !ann.Labels.IsSet(IsPredicate) {
+			return
+		}
+
+		name, params := ParseAsPredicate(v)
+		if name != "contains.file" {
+			return
+		}
```
Comment (rvantonder) [inline internal/search/query/types.go:351]: Ideally we'd cover the case for `repo:contains(file:README)` too but I'm fuzzy about how much extra trouble that is in this PR.

**Review by rvantonder — APPROVED** [review]
Nice, the same thought crossed my mind to effectively alias the `contains.file` part, but I actually tried to implement it in terms of the `repoHasFile` text search job (which didn't succeed when I tried to hack it because it was too annoying).

File: internal/search/query/types.go (hunk 4 above)
Comment (camdencheek) [reply internal/search/query/types.go:351]: I briefly looked at that, but the problem is things get messy trying to determine whether the predicate is handled by predicate expansion or whether it's handled by job construction. In this case, there is no clean split -- if `repo:contains()` has a content field, it's handled by expansion, otherwise it's handled during job construction. 

It's definitely possible, but I'll push that to another PR so we can evaluate the mess separately.

Comment (rvantonder) [reply internal/search/query/types.go:351]: ayup, no worries! I have an itch to help make this better too, definitely possible.

File: internal/search/query/predicate.go (hunk 1 above)
Comment (camdencheek) [reply internal/search/query/predicate.go:241]: > Do we have data on this?

No, not really. "small" and "large" are fuzzy and relative. And actually, in practice, there are probably cases where `repo:contains.file()` would be better as a post-filter (3) than a pre-filter (2), especially on an instance like `sourcegraph.com` where we have _many_ repos.

The question is whether it's cheaper to:
1) Have the backend do work that might be thrown away (filtered out)
2) Have the engine do work that might not actually save the backend any work (e.g. repo:contains.file(.*))

This is a tough question to answer, and we don't keep table statistics like real databases do, which means it's almost impossible for us to answer this dynamically, so the best we can do is optimize for the average case. That said, we haven't really experimented with post-filters, so we don't have any hard numbers to validate this, it's just a hunch.

Comment (tbliu98) [reply internal/search/query/predicate.go:241]: Appreciate the clarification. My plan is to shift focus to #30733 since we've identified it as a priority item for this quarter so grokking predicates is now very important to me

**Comment by keegancsmith** [comment]
> Note that this does not cover the repo:contains(file:xyz) case since
we do not currently have an efficient implementation of a "repo contains content"
filter.

I see this code is all in our job layer, but FYI for zoekt it should pretty much be the same way you are doing contains file queries. A query like `(type:repo (...))` or directly using `List` should work.

**Comment by camdencheek** [comment]
@keegancsmith yes! This PR was just aimed at unifying how we handle `repo:contains.file()` and `repohasfile:`. See the WIP PR [here](https://github.com/sourcegraph/sourcegraph/pull/38988) for a rewrite of all the `repo:contains...` logic that [uses zoekt directly](https://github.com/sourcegraph/sourcegraph/pull/38988/files#diff-625b5188ac3d506dbb4a20b5904654d7a8f6d221bc5ed829270829cdd0887f4bR614) for indexed revs.

**Comment by camdencheek** [comment]
FWIW, I spent a lot of time trying to get `(type:repo ...)` to work with `List`, but it doesn't seem to do what I want it to. I've resorted to a `StreamSearch` with `ShardMaxMatchCount: 1`. Basically, I think what's happening is we expand `(type:repo ...)` nodes [here](https://sourcegraph.com/github.com/sourcegraph/zoekt@e51248fb93d565f4f9b26ed40e7755342742c7b2/-/blob/shards/eval.go?L79-79), but then we still [run `List` with the expanded query](https://sourcegraph.com/github.com/sourcegraph/zoekt@e51248fb93d565f4f9b26ed40e7755342742c7b2/-/blob/shards/eval.go?L84-84), which ultimately [still runs a search on each repo in the shard](https://sourcegraph.com/github.com/sourcegraph/zoekt@e51248fb93d565f4f9b26ed40e7755342742c7b2/-/blob/eval.go?L579-582). Since the only nodes left in our query at this point are `RepoList` nodes and `BranchesRepos` nodes, it doesn't return any search results. Ideally, we should be able to just simplify a repo-only query tree those down to a single `RepoList` and return that, but that started getting complicated so I didn't continue down that path.
